2022/Q4/artifacts/PoS-pseudocode/PoS-model.md

```diff
@@ -301,34 +304,43 @@ func unbond(validator_address, delegator_address, total_amount)
 
 ```go
 /* COMMENT: Something to check for correctness https://github.com/informalsystems/partnership-heliax/pull/16#discussion_r924319213 */
-//This function is called by transactions tx_withdraw_unbonds_validator and tx_withdraw_unbonds_delegator
+// This function is called by transactions tx_withdraw_unbonds_validator and tx_withdraw_unbonds_delegator.
+// Withdraws all tokens from PoS for all unbonds between the delegator adn validator that are fully matured as of the current epoch. Applies slashes to the unbonds, transferring all slashed tokens to the slash pool address and all unslashed tokens back to the source delegator.
 func withdraw(validator_address, delegator_address)
 {
   var delunbonds = {<start,end,amount> | amount = unbonds[delegator_address][validator_address].deltas[(start, end)] > 0 && end <= cur_epoch }
-  //substract any pending slash before withdrawing
   forall (<start,end,amount> in selfunbonds) do
-    // set of slashes that happened while the bond associated with the unbond was contributing to the validator's stake
-    var set_slashes = {s | s in slashes[validator_address] && start <= slash.epoch && slash.epoch < end - unbonding_length }
+    // Apply slashes that happened while the bond associated with the unbond was contributing to the validator's stake
+    var set_slashes = {s | s in slashes[validator_address] && start <= slash.epoch && slash.epoch < end - unbonding_length - cubic_slash_window_width}
     var amount_after_slashing = compute_amount_after_slashing(set_slashes, amount)
     balance[delegator_address] += amount_after_slashing
-    balance[pos] -= amount_after_slashing
-    //remove unbond
+    balance[slash_pool] += amount - amount_after_slashing
```

Comment:
I believe we should only transfer tokens to the `slash_pool` address at the end of an epoch. The tokens being transferred here have already been transferred at the end of the epoch when the slash was processed.

Reply:
Not all of the tokens transferred here have been transferred during `end_of_epoch` I believe. During `end_of_epoch`, the token amount transferred is the additional amount of tokens to deduct from the deltas such that we achieve the desired total slash amount when considering already existing unbonds of bonds that contributed to the infraction.

The amount transferred during the withdrawal here is the amount of these already existing unbonds (which already reduced the deltas before processing the slashes) that need to be sent to the slash pool. Does this make sense?